sw/device/tests/otbn_mem_scramble_test.c

```diff
@@ -4,107 +4,216 @@
+ */
+static void otbn_write_mem_words(const otbn_t *ctx, const int num,
+                                 const uint32_t *word_addrs,
+                                 otbn_write_t otbn_write) {
+  for (int i = 0; i < num; ++i) {
+    otbn_write(&ctx->dif, word_addrs[i] * sizeof(uint32_t),
+               (void *)&word_addrs[i], sizeof(uint32_t));
+  }
 }
 
 /**
- * Check that the application is loaded correctly to the IMEM and DMEM.
+ * Check whether the contents at addresses `word_addrs` of an OTBN memory match
+ * the reference data pointed at `word_addrs`.
  *
  * @param ctx The otbn context object.
- * @param app The application to match with OTBN memory.
- * @param match_expected Indicates whether the checking is expected to match.
+ * @param num The number of addresses to check.
+ * @param word_addrs The word addresses to check.
+ * @param otbn_read Pointer to the function to read the memory. It can be
+ *                  either `dif_otbn_imem_read` or `dif_otbn_dmem_read`.
+ * @param num_matches Pointer to the number of observed matches.
+ * @param num_intg_errors Pointer to the number of observed integrity errors.
  */
-static void otbn_check_app(otbn_t *ctx, const otbn_app_t app,
-                           bool match_expected) {
-  const size_t imem_size = app.imem_end - app.imem_start;
-  const size_t data_size = app.dmem_data_end - app.dmem_data_start;
-  const uint8_t *imem_start = app.imem_start;
-  const uint8_t *dmem_start = app.dmem_data_start;
-
-  // Memory images and offsets must be multiples of 32b words.
-  CHECK(imem_size % sizeof(uint32_t) == 0);
-
-  // Check the IMEM content.
-  otbn_check_mem(ctx, imem_start, imem_size, match_expected,
-                 dif_otbn_imem_read);
-
-  // Check the DMEM content.
-  otbn_check_mem(ctx, dmem_start, data_size, match_expected,
-                 dif_otbn_dmem_read);
+static void otbn_check_mem_words(const otbn_t *ctx, const int num,
+                                 const uint32_t *word_addrs,
+                                 otbn_read_t otbn_read, int *num_matches,
+                                 int *num_intg_errors) {
+  *num_matches = 0;
+  *num_intg_errors = 0;
+
+  uint32_t word;
+  bool match;
+  for (int i = 0; i < num; ++i) {
+    // If the memory has been scrambled we expect to receive an IRQ due to the
+    // integrity error.
+    has_irq_fired = false;
+    otbn_read(&ctx->dif, word_addrs[i] * sizeof(uint32_t), (void *)&word,
+              sizeof(uint32_t));
+    if (word_addrs[i] == word) {
+      match = true;
+    } else {
+      match = false;
+    }
+    if (match) {
+      *num_matches += 1;
+    }
+    if (has_irq_fired) {
+      *num_intg_errors += 1;
+    }
+    // It is possible that the integrity bits after re-scrambling are still
+    // valid.
+    if ((match == false) && (has_irq_fired == false)) {
+      LOG_INFO(
+          "Mismatch without integrity error: Entry %i, address %u, "
```

Comment:
I would print the address and the data in hex `0x%x`, it seems more useful for me.

Reply:
Done